pkg/edgeview/src/copyfile.go

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) 2021 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+type copyFile struct {
+	TokenHash []byte   `json:"tokenHash"`
+	Name      string   `json:"name"`
+	Size      int64    `json:"size"`
+	Sha256    string   `json:"sha256"`
+	ModTsec   int64    `json:"modtsec"`
+}
+
+type fileCopyStatus struct {
+	gotFileInfo   bool
+	filename      string
+	fileSize      int64
+	fileHash      string
+	currSize      int64
+	modTime       time.Time
+	buf           []byte
+	f             *os.File
+}
+
+func runCopy(opt string) {
+	path := strings.SplitN(opt, "cp/", 2)
+	if len(path) != 2 {
+		fmt.Printf("cp needs a cp/ and path input\n")
+		return
+	}
+	file := path[1]
+	info, err := os.Stat(file)
+	if err != nil {
+		fmt.Printf("os stat error %v\n", err)
+		return
+	}
+	//fmt.Printf("file info %+v\n", info)
+	cfile := copyFile{
+		TokenHash: tokenHash16,
+		Name:      info.Name(),
+		Size:      info.Size(),
+		ModTsec:   info.ModTime().Unix(),
+		Sha256:    fmt.Sprintf("%x", getFileSha256(file)),
+	}
+	jbytes, err := json.Marshal(cfile)
+	if err != nil {
+		fmt.Printf("json marshal error %v\n", err)
+		return
+	}
+	_ = websocketConn.WriteMessage(websocket.BinaryMessage, jbytes)
+
+	// server side set
+	isSvrCopy = true
+	copyMsgChn = make(chan []byte)
+	//websocketConn.SetReadDeadline(time.Now().Add(60 * time.Second))
+	ahead := make(chan struct{})
+	done := make(chan struct{})
+	t := time.NewTimer(30 * time.Second)
+	readerRunning := true
+
+	go func() {
+		for {
+			select {
+			case message := <-copyMsgChn:
+				if !strings.Contains(string(message), startCopyMessage) {
+					fmt.Printf("webc read message. %s\n", string(message))
+					//websocketConn.SetReadDeadline(time.Time{})
+					readerRunning = false
+					if !isClosed(ahead) {
+						close(ahead)
+					}
+					isSvrCopy = false
+					return
+				} else {
+					//fmt.Printf("start file transfer\n")
+					close(ahead)
+				}
+
+			case <-t.C:
+				readerRunning = false
+				if !isClosed(ahead) {
+					close(ahead)
+				}
+				isSvrCopy = false
+				return
+
+			case <- done:
+				t.Stop()
+				isSvrCopy = false
+				return
+			}
+		}
+	}()
+
+	<- ahead
+	if !readerRunning {
+		return
+	}
+	f, err := os.Open(file)
+	if err != nil {
+		fmt.Printf("os open error %v\n", err)
+		return
+	}
+	defer f.Close()
+
+	buffer := make([]byte, 8192)
+	totalBytes := 0
+	for {
+		n, err := f.Read(buffer)
+		if err != nil {
+			fmt.Printf("file read error %v\n", err)
+			return
+		}
+		_ = websocketConn.WriteMessage(websocket.BinaryMessage, buffer[:n])
```

Comment:
Should we not check for error and terminate the file transfer? For example when the client sends RST.

Reply:
yes, we can do that.